api/api/migrations/0058_userpreferences.py

```diff
@@ -0,0 +1,34 @@
+# Generated by Django 4.2.10 on 2024-03-13 21:24
+
+from django.conf import settings
+from django.db import migrations, models
+import django.db.models.deletion
+
+
+def create_user_preferences(apps, schema_editor):
+    UserPreferences = apps.get_model("api", "UserPreferences")
+    User = apps.get_model("auth", "User")
+    for user in User.objects.all():
+        if not hasattr(user, "userpreferences"):
```

Comment:
Just a nit (and mostly a tip for the future, really), but we would ideally do this kind of filtering at the ORM level, so that it's executed in the DB, rather than in Python. Calling `.all()` on a model manager will load literally every single instance of the model into memory. That might not matter in this case, because I don't think we have many User instances in our live environments, but at some future date it could.

In Django ORM you could do that a couple of ways, in order of best to worst options. You can definitely use the first one in this case, but for more complex cases, the others might be necessary. For all of these we would still want to use QuerySet slices to avoid loading all selected instances into memory: even a subset could be huge. These alternative queries reduce the selected rows, but don't necessarily solve the problem of loading too many instances into memory all at once.

So for this case, you'd just use `isnull` to do an outer join:

```py
User.objects.filter(userpreferences__isnull=True)
```

Which turns into this SQL:

```sql
SELECT
    "auth_user"."id", "auth_user"."password", "auth_user"."last_login", "auth_user"."is_superuser", "auth_user"."username", "auth_user"."first_name", "auth_user"."last_name", "auth_user"."email", "auth_user"."is_staff", "auth_user"."is_active", "auth_user"."date_joined"
FROM
    "auth_user"
    LEFT OUTER JOIN
        "api_userpreferences" ON ("auth_user"."id" = "api_userpreferences"."user_id")
WHERE
    "api_userpreferences"."id" IS NULL
```

I don't think there's any reason _not_ to use that approach for this particular kind of one-to-one relationship.

**For other circumstances**, you can also do the following, to avoid filtering in Python.

With an inner-select and not-in (probably the simplest, but potentially also problematic):

```py
User.objects.exclude(pk__in=UserPreferences.objects.values_list("user", flat=True))
```

Which turns into this SQL:

```sql
SELECT
    "auth_user"."id", "auth_user"."password", "auth_user"."last_login", "auth_user"."is_superuser", "auth_user"."username", "auth_user"."first_name", "auth_user"."last_name", "auth_user"."email", "auth_user"."is_staff", "auth_user"."is_active", "auth_user"."date_joined"
FROM
    "auth_user"
WHERE NOT (
    "auth_user"."id" IN (SELECT U0."user_id" FROM "api_userpreferences" U0)
)
```

With a join:

```py
from django.db.models import Exists, OuterRef

User.objects.filter(Exists(UserPreferences.objects.filter(user=OuterRef("pk"))))
```

which turns into this SQL:

```sql
SELECT
    "auth_user"."id", "auth_user"."password", "auth_user"."last_login", "auth_user"."is_superuser", "auth_user"."username", "auth_user"."first_name", "auth_user"."last_name", "auth_user"."email", "auth_user"."is_staff", "auth_user"."is_active", "auth_user"."date_joined"
FROM
    "auth_user"
WHERE
    EXISTS (
        SELECT 1 AS "a" FROM "api_userpreferences" U0 WHERE U0."user_id" = ("auth_user"."id") LIMIT 1
    )
```

Any of these are ideally combined with QuerySet slicing to avoid loading the full list into memory. Any of these query sets can be used like so, to achieve chunking:

```py
qs = User.objects.whatever()
slice_size = 10

while (slice := qs[:slice_size]):
    for user in slice:
        # do whatever
```

That has the added benefit of being a noop if `qs` is empty. QueryString slices are already evaluated, so assigning it to `slice` saves any second query, and `bool(empty_qs)` evaluates to False, so the while loop combined with walrus operator works great here.

**But anyway, I don't think this will be a problem for this PR**. Each `ThrottledApplication` instance is _able_ to have a user (https://github.com/jazzband/django-oauth-toolkit/blob/master/oauth2_provider/models.py#L102-L108) but it looks like that isn't something the library automatically enforces. I went ahead and queried production API database to be sure:

```
deploy@localhost:openledger> select count(*) from api_throttledapplication where user_id is not null;
+-------+
| count |
|-------|
| 0     |
+-------+
SELECT 1
Time: 0.227s
deploy@localhost:openledger> select count(*) from api_throttledapplication where user_id is null;
+-------+
| count |
|-------|
| 5031  |
+-------+
SELECT 1
Time: 0.233s
```

And more succinctly:

```
deploy@localhost:openledger> select count(*) from auth_user;
+-------+
| count |
|-------|
| 2     |
+-------+
SELECT 1
Time: 0.227s
```

So anyway, definitely not a problem in this particular case, but just wanted to share how to avoid this in the future in case it ever would be a problem, and give examples of how it can be achieved. Django ORM is very powerful but terribly complex :cold_sweat:

Reply:
THANK YOU for taking the time to write this all out! I'll definitely be referencing it in the future 😄 I took the easiest approach here because I _knew_ we had very few user accounts at this point, so iterating over all of them would not be a problem. I appreciate the reminder about how this "giant hammer" approach could be problematic in the future!